artemis-protocols/artemis-mqtt-protocol/src/test/java/org/apache/activemq/artemis/protocol/mqtt/SessionStateManagerTest.java

```diff
@@ -0,0 +1,151 @@
+         MQTTSessionState sessionState = new MQTTSessionState(clientId, stateManager);
+         for (Pair<MqttTopicSubscription, Integer> sub : subs) {
+            sessionState.addSubscription(sub.getA(), MQTTUtil.MQTT_WILDCARD, sub.getB());
+         }
+      }
+
+      Wait.assertEquals(SESSION_COUNT, () -> server.locateQueue(MQTTUtil.MQTT_SESSION_STORE).getMessageCount(), 2000, 100);
+
+      server.stop();
+      server.start();
+
+      Wait.assertEquals(SESSION_COUNT, () -> server.locateQueue(MQTTUtil.MQTT_SESSION_STORE).getMessageCount(), 2000, 100);
+
+      stateManager = MQTTSessionStateManager.getInstance(server);
+
+      for (Pair<String, Collection<Pair<MqttTopicSubscription, Integer>>> session : sessions) {
+         String clientId = session.getA();
+         Collection<Pair<MqttTopicSubscription, Integer>> mySubs = session.getB();
+         Map<String, MQTTSessionState> sessionStates = stateManager.getSessionStates();
+         assertTrue(sessionStates.containsKey(clientId));
+         MQTTSessionState state = sessionStates.get(clientId);
+         Collection<Pair<MqttTopicSubscription, Integer>> brokerSubs = state.getSubscriptionsPlusID();
+         assertEquals(mySubs.size(), brokerSubs.size());
+         for (Pair<MqttTopicSubscription, Integer> mySub : mySubs) {
+            boolean found = false;
+            for (Pair<MqttTopicSubscription, Integer> brokerSub : brokerSubs) {
+               if (compareSubs(mySub.getA(), brokerSub.getA())) {
+                  found = true;
+                  assertEquals(mySub.getB(), brokerSub.getB());
+                  break;
+               }
+            }
+            assertTrue(found);
+         }
+      }
+   }
+
+   private boolean compareSubs(MqttTopicSubscription a, MqttTopicSubscription b) {
+      if (a == b)
+         return true;
+      if (a == null || b == null)
+         return false;
+      if (a.topicName() == null) {
+         if (b.topicName() != null)
+            return false;
+      } else if (!a.topicName().equals(b.topicName())) {
+         return false;
+      }
+      if (a.option() == null) {
+         if (b.option() != null)
+            return false;
+      } else {
+         if (a.option().qos() == null) {
+            if (b.option().qos() != null)
+               return false;
+         } else if (a.option().qos().value() != b.option().qos().value()) {
+            return false;
+         }
+         if (a.option().retainHandling() == null) {
+            if (b.option().retainHandling() != null)
+               return false;
+         } else if (a.option().retainHandling().value() != b.option().retainHandling().value()) {
+            return false;
+         }
+         if (a.option().isRetainAsPublished() != b.option().isRetainAsPublished()) {
+            return false;
+         }
+         if (a.option().isNoLocal() != b.option().isNoLocal()) {
+            return false;
+         }
```

Comment:
Its a bit weird mixing single-line-ifs with-braces and without-braces in the same method multiple times. With braces is nicer and safer. Some separating newlines would make it a lot more readable too.

(I find it bizarre the build enforces there is a space before the brace but also allows it not to be there hehe)